crates/datafusion/src/physical_plan/clickhouse.rs

```diff
@@ -85,6 +86,24 @@ impl FromStr for BuiltinScalarFunction {
     }
 }
 
+/// Implement like ad-hoc polymorphic function encapsulation
+/// select specific implementation functions based on data type
+/// return Err if it doesn't match
+macro_rules! poly_func_impl {
+    ($ARGS:expr, $SCHEMA:expr, $($DATA_TYPE:pat => $Func:ident$(,)?)*) => {{
+	match $ARGS[0].data_type($SCHEMA) {
+	    $( Ok($DATA_TYPE) => $Func, )*
+	    other => {
```

Comment:
your poly_func_impl is just pattern match wrapper.  But you just call this macro once. I, personally, hope that all friends to reduce the usage of macro. They are hard to debug when some thing wrong. If you feel ```return Err(...)``` is too verbose to reduce the readability, you just abstract it as a function call (with inline attr if you want).

Reply:
OK, subsequent use of macros will be reduced if possible.

Reply:
thanks :pray: abuse of macros can cause many problems. Higher-kinded constructions may be good alternatives, which may be discussed more by us in the future works :smile:

Reply:
Hi @pandaplusplus, at a fast glimpse of the code, It seems that the macro `poly_func_impl` can be incorporated into the `wrap_datetime_fn` by adding an optional schema argument? Such as:
```diff
Index: crates/datafusion/src/physical_plan/clickhouse.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/crates/datafusion/src/physical_plan/clickhouse.rs b/crates/datafusion/src/physical_plan/clickhouse.rs
--- a/crates/datafusion/src/physical_plan/clickhouse.rs	(revision a08e1ce18d7736d0c84766e0dbcdacc86228d470)
+++ b/crates/datafusion/src/physical_plan/clickhouse.rs	(date 1625490171287)
@@ -337,12 +337,12 @@
 macro_rules! wrap_datetime_fn {
     ( $(
         $(#[$OUTER:meta])* $NAME:literal => fn $FUNC:ident {
-            $( $DATA_TYPE:pat => fn $OP:ident($INPUT_TY:ty $(, $TZ:ident)? ) -> $OUTPUT_TY:ty, )*
+            $( $DATA_TYPE:pat( $( $SCHEMA:expr )? ) => fn $OP:ident($INPUT_TY:ty $(, $TZ:ident)? ) -> $OUTPUT_TY:ty, )*
         }
     )* ) => { $(
         $(#[$OUTER])*
         pub fn $FUNC(args: &[ColumnarValue]) -> $crate::error::Result<ColumnarValue> {
-            match args[0].data_type() {
+            match args[0].data_type( $( $SCHEMA )? ) {
                 $(
                     $DATA_TYPE => match &args[0] {
 			ColumnarValue::Array(array) => {
```

The main problem of the `wrap_datetime_fn` macro is that it does not accept multiple arguments. If you have interests, we can discuss on WeChat about how to improve the macros. DF has many beautiful and useful macros, maybe we can learn from them.

Reply:
@whjpji good, you can send out a PR. KISS is always we want.